Sources/Gui/SDLGLDevice.h

```diff
@@ -25,16 +25,54 @@
 
 #include <Draw/IGLDevice.h>
 
+#define SPADES_SDLGL_TEXTURE_UNITS 16
+
 namespace spades {
 	namespace gui {
+		//this probably shouldn't be here but im not too experienced with C++
+
+		typedef struct SDLGLState_s {
+			SDLGLState_s(): dm_b(true), cm_r(true), cm_g(true), cm_b(true), cm_a(true), ff_val(spades::draw::IGLDevice::CCW), df_func(spades::draw::IGLDevice::Less){}
+			//DepthRange
+			float dr_near, dr_far;
+			//Viewport
+			int vp_x, vp_y;
+			unsigned int vp_width, vp_height; //todo sizei
+			//ClearDepth
+			float cd_v;
+			//ClearColor
+			float cc_r, cc_g, cc_b, cc_a;
+			//Depthmask - initially true
+			bool dm_b;
+			//Colormask - initially true
+			bool cm_r, cm_g, cm_b, cm_a;
+			//FrontFace
+			spades::draw::IGLDevice::Enum ff_val;
+			//Linewidth
+			float lw_w;
+			//DepthFunc
+			spades::draw::IGLDevice::Enum df_func;
+			//UseProgram
+			spades::draw::IGLDevice::UInteger up_program;
+			//Texture stuff
+			unsigned char tex_active;
+			//change the number if you want more texture units!
+			spades::draw::IGLDevice::UInteger tex_id[SPADES_SDLGL_TEXTURE_UNITS];
+			spades::draw::IGLDevice::Enum tex_target[SPADES_SDLGL_TEXTURE_UNITS];
```

Comment:
I'm no expert of OpenGL, but it seems that OpenGL doesn't work like that:

https://www.khronos.org/opengl/wiki/Texture#Texture_image_units

> Each texture image unit supports bindings to all targets. So a 2D texture and an array texture can be bound to the same image unit, or different 2D textures can be bound in two different image units without affecting each other. So which texture gets used when rendering? In GLSL, this depends on the type of sampler that uses this texture image unit.